### PR TITLE
pipeline & basic framework

### PR DESCRIPTION
**Pull request by runningmaverick**

### CODE REVIEW

File: cmd/ceph/main.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+import (
+	"fmt"
+	"github.com/ceph/go-ceph/rados"
+	"io/ioutil"
+)
+
+func main() {
+	mons := "172.11.51.7:6789"
+	secret := []byte("AQB+OJ9YAJSoHhAACOtovk+EY2QDiEksGM1XkA==")
```
Comment (xiaoyunwu) [inline cmd/ceph/main.go]: todo: change this to command line parameter that can be set in the launch

File: cmd/ceph/main.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
+package main
+
+import (
+	"fmt"
+	"github.com/ceph/go-ceph/rados"
```
Comment (xiang90) [inline cmd/ceph/main.go]: separate internal and extern imports

File: compute/pb/compute.proto [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+message ReduceIn {
+    repeated KV kvs = 1;
+}
+
+message ReduceOut {
+    repeated KV kvs = 1;
+}
+
+service Compute {
```
Comment (xiang90) [inline compute/pb/compute.proto]: move service definition to the top.

File: cmd/ceph/main.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+	"io/ioutil"
+)
+
+func main() {
+	mons := "172.11.51.7:6789"
```
Comment (xiang90) [inline cmd/ceph/main.go]: this should not be a inline const.

File: cmd/ceph/main.go [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+	secret := []byte("AQB+OJ9YAJSoHhAACOtovk+EY2QDiEksGM1XkA==")
+
+	conn, _ := rados.NewConn()
+	err := ioutil.WriteFile("ceph-keyfile", secret, 0600)
+	if err != nil {
+		fmt.Println("Failed to store keyfile:", err)
```
Comment (xiang90) [inline cmd/ceph/main.go]: use log.fatalf

File: cmd/ceph/main.go [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+		return
+	}
+
+	args := []string{"--mon-host", mons, "--keyfile", "ceph-keyfile"}
+	err = conn.ParseCmdLineArgs(args)
+	if err != nil {
+		fmt.Println("Failed to Parse args:", err)
```
Comment (xiang90) [inline cmd/ceph/main.go]: same here

File: cmd/ceph/main.go [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+		return
+	}
+
+	err = conn.Connect()
+	if err != nil {
+		fmt.Println("Failed to connect:", err)
```
Comment (xiang90) [inline cmd/ceph/main.go]: same here.

File: cmd/ceph/main.go [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+		return
+	}
+
+	ioctx, err := conn.OpenIOContext("kmr")
+	if err != nil {
+		fmt.Println(fmt.Sprintf("Cannot open %s:", "kmr"), err)
```
Comment (xiang90) [inline cmd/ceph/main.go]: fatalf

File: cmd/ceph/main.go [hunk 9]
```diff
@@ -0,0 +1,63 @@
+package main
```
Comment (xiang90) [inline cmd/ceph/main.go]: is this a benchmark tool? or a testutil?

File: cmd/mapper/main.go [hunk 10]
```diff
@@ -0,0 +1,29 @@
+package main
+
+import (
+	"flag"
+	"os"
+
+	"github.com/ceph/go-ceph/rados"
+	"github.com/naturali/NiServer/pkg/log"
```
Comment (xiang90) [inline cmd/mapper/main.go]: use something like https://github.com/sirupsen/logrus?

File: compute/pb/compute.proto [hunk 11]
```diff
@@ -0,0 +1,29 @@
+syntax = "proto3";
+
+package pb;
+
+message KV {
```
Comment (xiang90) [inline compute/pb/compute.proto]: please add comments on these messages and service.

**Comment by xiang90** [comment]
From high level, both the mapper and reducer should be decoupled with ceph. Ceph should be a simple IO provider that implements an IO interface or something. You can do things like

mapper --ioprovider=ceph  for example.

File: cmd/mapper/main.go [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+
+	"github.com/naturali/kmr/job"
+)
+
+func main() {
+	fs := flag.NewFlagSet("name", flag.ExitOnError)
```
Comment (xiang90) [inline cmd/mapper/main.go]: why do you create a new flag set here?

File: job/job.go [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+
+type Context struct {
+}
+
+type JobInfo struct {
+	MapperCount  int
+	ReducerCount int
+}
+
+type Job interface {
```
Comment (xiang90) [inline job/job.go]: comments on the public interface?

File: job/job.go [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+import (
+	"fmt"
+
+	"github.com/naturali/kmr/records"
+)
+
+type Context struct {
+}
+
+type JobInfo struct {
```
Comment (xiang90) [inline job/job.go]: JobInfo -> Config

File: job/job.go [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+	}
+}
+
+type Task struct {
+	id string
+}
+
+func (job *kubeJob) Launch() {
+	// fake inits
+	dataSource = []string{"key1", "key2"}
```
Comment (xiang90) [inline job/job.go]: will this even compile? where do you define dataSource?

File: cmd/ceph/main.go (hunk 9 above)
Comment (runningmaverick) [inline cmd/ceph/main.go]: it's a testutil

**Comment by comzyh** [comment]
This PR is outdated and merged.

**Pull request closed without merging**